### PR TITLE
pkg/semtech-loramac: use internal eeprom to store lorawan config (deveui, appeui, etc)

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PRs use the new `periph_eeprom` in the semtech-loramac package to automatically read lorawan configuration from the internal eeprom, if the cpu provides this feature.

For example, using the `tests/pkg_semtech-loramac` application, one just has to set the required EUI and Keys, then type `loramac save` in the shell. After a reboot, no need to reenter the keys, they are automatically re-read from eeprom.

The automatic read uses an internal magic number that is stored at a certain position of the EEPROM. These are defines and can be changed at build time.

It's also possible to erase previously store configuration parameters.

Finally, the frame counters can also be stored and reused after a reboot. This is specific to ABP activation that needs to keep this value after a reboot, otherwise this needs to be cleared manually in the network backend.

Note that I add to modify the module dependency of `periph_eeprom` when it's loaded as an optional dependency on STM32.

### Issues/PRs references

None

### CODE REVIEW

**Comment by jia200x** [comment]
this is awesome!

File: pkg/semtech-loramac/contrib/semtech_loramac.c [hunk 1]
```diff
@@ -708,3 +715,93 @@ uint8_t semtech_loramac_recv(semtech_loramac_t *mac)
...
+    eeprom_read(pos, magic, SEMTECH_LORAMAC_EEPROM_MAGIC_LEN);
+    if (strcmp((char*)magic, "RIOT") != 0) {
+        DEBUG("[semtech-loramac] no configuration present on EEPROM\n");
+        return;
+    }
+
+    size_t end = (pos + SEMTECH_LORAMAC_EEPROM_MAGIC_LEN +
+                  LORAMAC_DEVEUI_LEN + LORAMAC_APPEUI_LEN +
+                  LORAMAC_APPKEY_LEN + LORAMAC_APPSKEY_LEN +
+                  LORAMAC_NWKSKEY_LEN + LORAMAC_DEVADDR_LEN + 4);
```
Comment (jia200x) [inline pkg/semtech-loramac/contrib/semtech_loramac.c]: what does this `4` in the end represents?

Comment (aabadie) [reply pkg/semtech-loramac/contrib/semtech_loramac.c]: These are the 4 bytes used for the uplink counter (uint32_t)

Comment (jia200x) [reply pkg/semtech-loramac/contrib/semtech_loramac.c]: maybe define it as a LORAMAC_UPCOUNTER_LEN?

File: pkg/semtech-loramac/contrib/semtech_loramac.c [hunk 2]
```diff
@@ -708,3 +715,93 @@ uint8_t semtech_loramac_recv(semtech_loramac_t *mac)
...
+}
+
+void _set_uplink_counter(semtech_loramac_t *mac, uint8_t *counter)
+{
+    DEBUG("[semtech-loramac] uplink counter: %lu\n", *(uint32_t *)counter);
+
+    mutex_lock(&mac->lock);
+    MibRequestConfirm_t mibReq;
+    mibReq.Type = MIB_UPLINK_COUNTER;
+    mibReq.Param.UpLinkCounter = *(uint32_t *)counter;
```
Comment (jia200x) [inline pkg/semtech-loramac/contrib/semtech_loramac.c]: this breaks the strict aliasing rule (as seen in Murdock output). Also the way how this variable is read depends on the endianess of the CPU.

**Comment by jia200x** [comment]
@aabadie works as expected. There's only a minor thing: the ABP counter should be stored always in EEPROM (if available). It's part of the LoRaWAN specification.

We could maybe provide the persistent ABP counters in another PR (e.g init for restore and send for save) and remove that part from this PR. What do you think?

**Comment by ZetaR60** [comment]
Very nice.

@aabadie I am concerned that using a macro to define the eeprom address will become awkward when there are multiple things using eeprom. What do you think about an eeprom registry?

**Comment by ZetaR60** [comment]
@aabadie I have created a rough draft of a simple EEPROM registry in #9537. I followed the general method you used here of a location + magic number.

**Comment by aabadie** [comment]
> the ABP counter should be stored always in EEPROM (if available). It's part of the LoRaWAN specification.

You are right @jia200x . Here I just did it in a safe way for the eeprom write cycles. I can add it later in this PR (because I'm now on vacation until the beginning of august).

> What do you think about an eeprom registry?

It sounds interesting. The macro defines are obviously not perfect but simple :)
I won't have the time to look in detail at #9537 but I like the idea in general and will be happy to reuse it in this PR when you think it's ready.

**Comment by aabadie** [comment]
> the ABP counter should be stored always in EEPROM (if available). It's part of the LoRaWAN specification.

@jia200x, I pushed a couple of commits for adding this. It works with ABP after rebooting without having to save the config before reboot.

**Comment by jia200x** [comment]
@aabadie I will try to find some time between today and tomorrow to check again

**Comment by aabadie** [comment]
Great thanks @jia200x !

**Comment by aabadie** [comment]
@jia200x, could find some time for testing ?

**Comment by jia200x** [comment]
yes, it freezes with abp :(. Since #8864 got merged, could you rebase?

**Comment by aabadie** [comment]
> could you rebase?

it is already rebased. Maybe something went wrong. I'll check.

**Comment by jia200x** [comment]
@aabadie any news here? :)

**Comment by aabadie** [comment]
> it freezes with abp :(.

@jia200x, can you give more details about your setup, your lorawan parameters (datarate, confirmable/unconfirmable tx, etc) ?

I re-tested this PR using ABP activation against TTN and if you follow the recommendation regarding TTN setup, it works well with DR5, DR0 and using uncnf/cnf messages.
I suspect that you forgot to set RX2 DR to 3 before sending. If I do this it hangs indeed. But this is totally out of the scope of this PR.

**Comment by jia200x** [comment]
> I suspect that you forgot to set RX2 DR to 3 before sending. If I don't do this it hangs indeed. But this is totally out of the scope of this PR.

I will try it again, but I was using uncnf messages and got no response from the shell.

**Comment by aabadie** [comment]
> I was using uncnf messages and got no response from the shell.

This sounds weird. Maybe an issue with RX window timings like the one reported @leandrolanzieri ? We could increase the value back to 50ms (it was shorten to 22ms in #8864)

In the meantime I pushed a commit that should fix the build on Murdock (but not the static tests)

**Comment by leandrolanzieri** [comment]
I tested this increasing the timeout to 50ms and get the same results as @jia200x:
```
> loramac set rx2_dr 3
2018-10-12 15:02:18,854 - INFO #  loramac set rx2_dr 3
> loramac set dr 0
2018-10-12 15:02:22,381 - INFO #  loramac set dr 0
> loramac tx RIOT uncnf
2018-10-12 15:02:31,240 - INFO #  loramac tx RIOT uncnf
2018-10-12 15:02:31,243 - INFO # [semtech-loramac] loramac cmd msg
2018-10-12 15:02:31,246 - INFO # [semtech-loramac] send frame RIOT
2018-10-12 15:02:31,250 - INFO # [semtech-loramac] MCPS request: unconfirmed TX
2018-10-12 15:02:31,258 - INFO # [semtech-loramac] MCPS request: OK
2018-10-12 15:02:32,210 - INFO # [semtech-loramac] MAC timer timeout
2018-10-12 15:02:33,166 - INFO # [semtech-loramac] MAC timer timeout
2018-10-12 15:02:34,121 - INFO # [semtech-loramac] MAC timer timeout
2018-10-12 15:02:35,077 - INFO # [semtech-loramac] MAC timer timeout
2018-10-12 15:02:36,031 - INFO # [semtech-loramac] MAC timer timeout
2018-10-12 15:02:36,987 - INFO # [semtech-loramac] MAC timer timeout
2018-10-12 15:02:37,943 - INFO # [semtech-loramac] MAC timer timeout
2018-10-12 15:02:38,899 - INFO # [semtech-loramac] MAC timer timeout
2018-10-12 15:02:39,854 - INFO # [semtech-loramac] MAC timer timeout
2018-10-12 15:02:40,810 - INFO # [semtech-loramac] MAC timer timeout
2018-10-12 15:02:41,765 - INFO # [semtech-loramac] MAC timer timeout
2018-10-12 15:02:42,720 - INFO # [semtech-loramac] MAC timer timeout
```

**Comment by aabadie** [comment]
Do you have the same problem with faster datarates (3,4,5) ? Do you see the messages passing on the gateway, LoRaWAN server ?

On my side, I'm testing with ST b-l072z-lrwan1 and it's working fine. I'll try to test with nucleo-l152re and an mbed shield next week.

**Comment by leandrolanzieri** [comment]
> Ok, thanks. What hardware are you using ? What lorawan provider TTN ?

I'm sorry, please ignore my last comment. I changed to the `lobaro-lorabox` board and forgot it has an SX1272.
Now I can send using ABP. So it is working for me.

**Comment by aabadie** [comment]
Great @leandrolanzieri !

**Comment by leandrolanzieri** [comment]
Tested on lobaro-lorabox board (SX1272):
- Entered key and EUIs for otaa and saved. Then turned the device off and on. Joined using otaa with the stored keys.
- Entered keys and address for abp. Joined using abp. Sent first message. Turned the device off and on. Joined again using abp and sent second message. The frame counters were correctly restored.

**Comment by jia200x** [comment]
@aabadie just one minor thing. Could you reflect this change in the README.md file? And also please squash ;).

Then we are ready to go!

**Comment by aabadie** [comment]
>  Could you reflect this change in the README.md file? And also please squash ;).

Will do

**Comment by aabadie** [comment]
@jia200x, I added 2 items in the README about save and erase subcommands. I also pushed an extra commit that reshape a bit the mardown formatting (my IDE linter was getting crazy). The result can be seen [here](https://github.com/aabadie/RIOT/tree/pr/pkg/semtech-loramac_eeprom/tests/pkg_semtech-loramac).

**Comment by jia200x** [comment]
> @jia200x, I added 2 items in the README about save and erase subcommands. I also pushed an extra commit that reshape a bit the mardown formatting (my IDE linter was getting crazy). The result can be seen here.

Looks good! Thank you.

**Comment by jia200x** [comment]
let's wait for Murdock

**Comment by jia200x** [comment]
...& GO

**Comment by jia200x** [comment]
@aabadie thank you so much for this contribution!

**Comment by aabadie** [comment]
> thank you so much for this contribution!

This will save us a lot of pain when playing with the shell and LoRaWAN :)

**Pull request merged**